### PR TITLE
[Pods] The pod UI-specific views

### PR DESCRIPTION
**Pull request by wavesoft**
image

### CODE REVIEW

**Comment by wavesoft** [comment]
Ready for review :)

File: src/js/components/PodDetail.js [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+      }
+    ];
+
+    METHODS_TO_BIND.forEach((method) => {
+      this[method] = this[method].bind(this);
+    });
+  }
+
+  componentWillMount() {
+    if (this.props.mode === 'pod') {
```
Comment (Poltergeist) [inline src/js/components/PodDetail.js]: Are you doing anything here? And is there anything which you have to do specifically in the `componentWillMount` function which you can't do in `componentDidMount`?

File: src/js/components/PodDetail.js [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    });
+  }
+
+  getRestartConfirmDialog() {
+    const {pod} = this.props;
+    const {state} = this;
+
+    return (
+      <Confirm
+        disabled={state.disabledDialog === ServiceActionItem.RESTART}
```
Comment (Poltergeist) [inline src/js/components/PodDetail.js]: personal #nit ... put this condition into a var like `isDisabled` next line as well.

I know that I am also not doing this always.

File: src/js/components/PodDetail.js [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    return (
+      <div className="flex-container-col">
+        <div className="container-pod
+          container-pod-divider-bottom-align-right
+          container-pod-short-top flush-bottom flush-top
+          service-detail-header media-object-spacing-wrapper
+          media-object-spacing-narrow">
+          <Breadcrumbs />
+          <PodInfo onActionsItemSelection={this.onActionsItemSelection}
+            pod={pod} tabs={this.tabs_getUnroutedTabs()} />
```
Comment (Poltergeist) [inline src/js/components/PodDetail.js]: Please place each prop on its own line.

File: src/js/components/PodDetailInstancesTab.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    super(...arguments);
+
+    this.store_listeners = [
+      {name: 'state', events: ['success']}
+    ];
+  }
+
+  render() {
+    return (
+      <PodView pod={this.props.pod} inverseStyle={true}
```
Comment (Poltergeist) [inline src/js/components/PodDetailInstancesTab.js]: Each prop on its own line please. So tagName and the first prop can be on the same line.

File: src/js/components/PodInfo.js [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+import HealthBar from './HealthBar';
+import PageHeader from './PageHeader';
+import Pod from '../structs/Pod';
+import ServiceActionItem from '../constants/ServiceActionItem';
+import StatusMapping from '../constants/StatusMapping';
+import StringUtil from '../utils/StringUtil';
+
+class PodInfo extends React.Component {
+  constructor() {
+    super();
```
Comment (Poltergeist) [inline src/js/components/PodInfo.js]: Maybe we want to pass the props ( arguments ) here

```JS
constructor(props) {
  super(props);
}
```

File: src/js/components/PodView.js [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+          let containerSpec = spec.getContainerSpec(container.name);
+          Object.keys(containerSpec.resources).forEach(function (key) {
+            resourcesSummary[key] = containerSpec.resources[key];
+          });
+
+          return {
+            id: container.getId(),
+            name: container.getName(),
+            status: container.getContainerStatus(),
+            address: container.endpoints.reduce(function (components, ep) {
```
Comment (Poltergeist) [inline src/js/components/PodView.js]: Lets have this in a separate variable.

File: src/js/components/PodView.js [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+              components.push(
+                <a className="text-muted"
+                  href={
+                    'http://' +
```
Comment (Poltergeist) [inline src/js/components/PodView.js]: Let us use a template string: 

```JS
href={`http://${instance.getAgentAddress()}:${ep.allocatedHostPort}`}

File: src/js/components/PodView.js [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+    let classes = classNames('pod-history-container-id is-expandable', {
+      'is-expanded': rowOptions.isExpanded
+    });
+
+    return this.renderWithClickHandler(rowOptions, classes, (
+      <CollapsingString string={row.id} />
+    ));
+  }
+
+  renderColumn_Address(prop, row, rowOptions = {}) {
```
Comment (Poltergeist) [inline src/js/components/PodView.js]: If there is not any magic behind this function name which I do not know, please do not use `_` underscores in function names.

**Comment by Poltergeist** [comment]
@wavesoft I added some comments I hope they can help you.

File: src/js/components/PodInfo.js (hunk 5 above)
Comment (mlunoe) [inline src/js/components/PodInfo.js]:
```js
constructor(props) {
  super(...arguments);
}
```

File: src/js/components/PodDetail.js (hunk 1 above)
Comment (wavesoft) [inline src/js/components/PodDetail.js]: Oops, you are right. Leftover of previous logic. Removed.

File: src/js/components/PodDetail.js (hunk 2 above)
Comment (wavesoft) [inline src/js/components/PodDetail.js]: 👍 this was copy-paste from the the TaskView, but spot-on.
Will do.

File: src/js/components/PodDetail.js (hunk 3 above)
Comment (wavesoft) [inline src/js/components/PodDetail.js]: 👍

File: src/js/components/PodDetailInstancesTab.js (hunk 4 above)
Comment (wavesoft) [inline src/js/components/PodDetailInstancesTab.js]: 👍

File: src/js/components/PodView.js (hunk 6 above)
Comment (wavesoft) [inline src/js/components/PodView.js]: Done

File: src/js/components/PodView.js (hunk 7 above)
Comment (wavesoft) [inline src/js/components/PodView.js]: Blast, good point! I am not so much used on template strings...

File: src/js/components/PodView.js (hunk 8 above)
Comment (wavesoft) [inline src/js/components/PodView.js]: No magic. I tend to avoid magic at any cost 😄 It was just a way to visually distinguish the different fragments... Renamed it.

File: src/js/components/PodDetail.js [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+  }
+
+  componentWillMount() {
+    if (this.props.mode === 'pod') {
+
+    }
+  }
+
+  componentDidMount() {
+    super.componentDidMount(...arguments);
```
Comment (mlunoe) [inline src/js/components/PodDetail.js]: No need to implement these if you are not going to do anything here

File: src/js/components/PodDetail.js [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+  }
+
+  componentWillUpdate() {
+    super.componentWillUpdate(...arguments);
```
Comment (mlunoe) [inline src/js/components/PodDetail.js]: Same here

File: src/js/components/PodDetail.js [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+    this.setState({
+      disabledDialog: null,
+      errorMsg: null,
+      serviceActionDialog: null
+    });
+  }
+
+  getRestartConfirmDialog() {
+    const {pod} = this.props;
+    const {state} = this;
```
Comment (mlunoe) [inline src/js/components/PodDetail.js]: const {serviceActionDialog , disabledDialog} = this.state;

File: src/js/components/PodDetail.js [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+    );
+  }
+
+  render() {
+    const {pod} = this.props;
+    let {serviceActionDialog} = this.state;
+
+    return (
+      <div className="flex-container-col">
+        <div className="container-pod
```
Comment (mlunoe) [inline src/js/components/PodDetail.js]: Let's get this into one line

File: src/js/components/PodDetail.js [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+    );
+  }
+
+  renderConfigurationTabView() {
+    return (
+      <ServiceDetailConfigurationTab service={this.props.pod} />
+    );
+  }
+
+  renderDebugTabView() {
```
Comment (mlunoe) [inline src/js/components/PodDetail.js]: Let's get these tabs routed

File: src/js/components/PodInfo.js (hunk 5 above)
Comment (mlunoe) [inline src/js/components/PodInfo.js]:
```js
constructor() {
  super(...arguments);
}
```

File: src/js/components/PodInfo.js [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+        buttonClassName="button button-stroke button-inverse dropdown-toggle"
+        dropdownMenuClassName="dropdown-menu"
+        dropdownMenuListClassName="dropdown-menu-list"
+        dropdownMenuListItemClassName="clickable"
+        wrapperClassName="dropdown flush-bottom"
+        items={dropdownItems}
+        persistentID={ServiceActionItem.MORE}
+        onItemSelection={this.props.onActionsItemSelection}
+        scrollContainer=".gm-scroll-view"
+        scrollContainerParentSelector=".gm-prevented"
```
Comment (mlunoe) [inline src/js/components/PodInfo.js]: Let's not add these classes!
If you don't want Gemini then use the `useGemini` option. But why would you want to disable it?

File: src/js/components/PodInfo.js [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+      return (
+        <span className={item.classes} key={index}>
+          {item.label}
+        </span>
+      );
+    });
+  }
+
+  render() {
+    let pod = this.props.pod;
```
Comment (mlunoe) [inline src/js/components/PodInfo.js]: #nit `let {pod} = this.props;`

File: src/js/components/PodView.js [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+  'renderColumn_ID',
+  'renderColumn_Resource',
+  'renderColumn_Status',
+  'renderColumn_Updated',
+  'renderColumn_Version'
+];
+
+class PodView extends mixin(SaveStateMixin, StoreMixin) {
+  constructor() {
+    super();
```
Comment (mlunoe) [inline src/js/components/PodView.js]: `...arguments`

File: src/js/components/PodView.js [hunk 17]
```diff
@@ -0,0 +1,370 @@
...
+      filter: {
+        text: '',
+        status: 'active'
+      }
+    };
+
+    this.saveState_properties = [
+    ];
+
+    this.store_listeners = [
```
Comment (mlunoe) [inline src/js/components/PodView.js]: If there are no store listeners, let's not add the StoreMixin

File: src/js/components/PodView.js (hunk 6 above)
Comment (mlunoe) [inline src/js/components/PodView.js]: There are a lot of nested loops here, which makes this hard to follow.
Could we break it up a bit

File: src/js/components/PodView.js [hunk 18]
```diff
@@ -0,0 +1,370 @@
...
+        updated: instance.getLastUpdated(),
+        status: instance.getInstanceStatus(),
+        version: spec.getVersion(),
+        children
+      };
+    });
+
+  }
+
+  getInstanceFilterStatus(instance) {
```
Comment (mlunoe) [inline src/js/components/PodView.js]: Couldn't we have this as a method on the struct

File: src/js/components/PodView.js [hunk 19]
```diff
@@ -0,0 +1,370 @@
...
+
+  renderWithClickHandler(rowOptions, className, content) {
+    return (
+        <div onClick={rowOptions.clickHandler} className={className}>
+          {content}
+        </div>
+      );
+  }
+
+  renderColumn_ID(prop, row, rowOptions = {}) {
```
Comment (mlunoe) [inline src/js/components/PodView.js]: We usually would do `renderColumnID`

File: src/js/components/PodViewFilter.js [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+
+const METHODS_TO_BIND = [
+  'getFilterButtonContent',
+  'handleStatusFilterChange',
+  'handleSearchStringChange'
+];
+
+class PodViewFilter extends React.Component {
+  constructor() {
+    super();
```
Comment (mlunoe) [inline src/js/components/PodViewFilter.js]: `...arguments`

File: src/js/components/PodViewFilter.js [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+
+  getChildrenArray() {
+    let {children} = this.props;
+    if (!children) {
+      return [];
+    }
+    if (!Array.isArray(children)) {
+      return [children];
+    }
+    return children;
```
Comment (mlunoe) [inline src/js/components/PodViewFilter.js]: #nit newline before this `return`

File: src/js/components/PodViewFilter.js [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+          inverseStyle={inverseStyle}
+          itemList={items.map(statusMapper)}
+          selectedFilter={filter.status} />
+      );
+  }
+
+  render() {
+    let {filter, inverseStyle} = this.props;
+    let children = this.getChildrenArray();
+    return (
```
Comment (mlunoe) [inline src/js/components/PodViewFilter.js]: #nit newline before this `return`

File: src/styles/components/pods-view.less [hunk 23]
```diff
@@ -0,0 +1,76 @@
+.pod-history-table {
+
+  tbody {
+    line-height: 1.8;
```
Comment (mlunoe) [inline src/styles/components/pods-view.less]: Let's use the canvas variables for line height definitions:
https://github.com/mesosphere/canvas/blob/master/styles/variables/variables-tables.less#L2269

NB: The reference is to a newer canvas then what we are currently using in dcos-ui!

**Comment by mlunoe** [comment]
Gave this a "quick" overlook, however, this PR is big! and that makes it much harder to review. Could you break this up into PR chunks of ~300 lines a piece?

File: src/js/components/PodView.js (hunk 18 above)
Comment (orlandohohmeier) [inline src/js/components/PodView.js]: @mlunoe this is a view specific helper and we shouldn't add those to the structs, but it could be useful to move it to a util.

File: src/js/components/PodDetail.js (hunk 13 above)
Comment (orlandohohmeier) [inline src/js/components/PodDetail.js]: Let's address this in a different PR, , as none of the service detail tabs is routed at the moment.

**Comment by wavesoft** [comment]
Ok, I will try to break this down into smaller PRs while addressing the comments.

File: src/js/components/PodInfo.js (hunk 14 above)
Comment (jfurrow) [inline src/js/components/PodInfo.js]: These are actually DOM node selectors used by the `Dropdown` component to determine which DOM node to attach event listeners to. So internally, the `Dropdown` will attach the scroll listener to the nearest `.gm-scroll-view` unless its parent matches the `.gm-prevented` selector.

**Comment by wavesoft** [comment]
This PR is broken down into the following PRs and is now closed:

#1067 , #1068 , #1069 , #1071 , #1077 , #1078 , #1079

File: src/js/components/PodDetail.js (hunk 13 above)
Comment (mlunoe) [inline src/js/components/PodDetail.js]: Thi is sad. Creating more tech debt in stead of doing it up in front. We have tasks to pay these down, but this will only get bigger.
I am happy with separate PRs as long as it gets handle with this task and not another task
